### PR TITLE
csi: CreateVolume and DeleteVolume support

### PR DESCRIPTION
**Pull request by lpabon**

### CODE REVIEW

File: csi/controller.go [hunk 1]
```diff
@@ -315,11 +316,152 @@ func osdVolumeAttributes(v *api.Volume) map[string]string {
...
+		}
+	} else {
+		// Get Capabilities and Size
+		spec.Size = req.GetCapacityRange().GetRequiredBytes()
+		for _, cap := range req.GetVolumeCapabilities() {
+			// Check access mode is setup correctly
+			mode := cap.GetAccessMode().GetMode()
+			if mode == csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER ||
+				mode == csi.VolumeCapability_AccessMode_MULTI_NODE_SINGLE_WRITER {
+				spec.Shared = true
```
Comment (piyush-nimbalkar) [inline csi/controller.go]: we can break here

File: csi/controller.go [hunk 2]
```diff
@@ -315,11 +316,152 @@ func osdVolumeAttributes(v *api.Volume) map[string]string {
...
+			req.GetVolumeId(),
+			err.Error())
+		dlog.Errorln(e)
+		return nil, status.Error(codes.Internal, e)
+	}
+
+	return &csi.DeleteVolumeResponse{
+		Reply: &csi.DeleteVolumeResponse_Result_{
+			Result: &csi.DeleteVolumeResponse_Result{},
+		},
```
Comment (piyush-nimbalkar) [inline csi/controller.go]: is the response supposed to be empty?

File: pkg/util/volume_test.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	"testing"
+
+	"github.com/libopenstorage/openstorage/api"
+	"github.com/libopenstorage/openstorage/volume/drivers/mock"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestVolumeFromNameFailedToLocateDueToTooManyVolumes(t *testing.T) {
```
Comment (piyush-nimbalkar) [inline pkg/util/volume_test.go]: should we add tests for inspect returning more than one volume or zero volumes (without error), although it does not happen?
similarly one for enumerate returning zero volumes without error?

File: csi/controller_test.go [hunk 4]
```diff
@@ -911,3 +911,222 @@ func TestControllerListVolumes(t *testing.T) {
...
+
+	// Setup mock
+	s.MockDriver().EXPECT().Delete(myid).Return(fmt.Errorf("TEST")).Times(1)
+
+	_, err := c.DeleteVolume(context.Background(), req)
+	assert.NotNil(t, err)
+	serverError, ok := status.FromError(err)
+	assert.True(t, ok)
+	assert.Equal(t, serverError.Code(), codes.Internal)
+	assert.Contains(t, serverError.Message(), "Unable to delete")
```
Comment (piyush-nimbalkar) [inline csi/controller_test.go]: nit: should we check for "TEST" to be sure the error is coming because of the delete call failing? in case, we are using "unable to delete" in multiple places.

File: csi/controller_test.go [hunk 5]
```diff
@@ -911,3 +911,222 @@ func TestControllerListVolumes(t *testing.T) {
 	}
 	assert.Equal(t, found, len(mockVolumeList))
 }
+
+func TestControllerCreateVolume(t *testing.T) {
```
Comment (piyush-nimbalkar) [inline csi/controller_test.go]: we could add some negative test cases for create volume

File: csi/controller.go (hunk 1 above)
Comment (lpabon) [reply csi/controller.go]: Good catch, but if in the future we add more settings, having a break there will have a bad side-effect. Having it go through the whole thing will provide easier enhancement in the near future (since this code and the spec are still heavily in flux).

File: csi/controller.go (hunk 2 above)
Comment (lpabon) [reply csi/controller.go]: Yeah, according to the spec. I found that really weird also :)

File: csi/controller_test.go (hunk 5 above)
Comment (lpabon) [reply csi/controller_test.go]: Ah! I forgot. You got it.

File: csi/controller_test.go (hunk 4 above)
Comment (lpabon) [reply csi/controller_test.go]: 👍

File: pkg/util/volume_test.go (hunk 3 above)
Comment (lpabon) [reply pkg/util/volume_test.go]: yeah, good point. I will add.

File: csi/controller.go (hunk 1 above)
Comment (piyush-nimbalkar) [reply csi/controller.go]: That is why we have tests :)
But I am fine with not adding the break as this code is changing and it does not have a very bad impact if we don't have the break
If you don't see this changing it in near future, then you have add it, else it's fine.

**Comment by lpabon** [comment]
still working on the unit tests.. do not review yet.

File: csi/controller.go (hunk 1 above)
Comment (lpabon) [reply csi/controller.go]: ha! yeah, you are correct. as long as we add tests with each change. You know, I agree, and I will add the change 👍

**Comment by lpabon** [comment]
All set. please review @piyush-nimbalkar @adityadani

File: pkg/util/volume_test.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+	// Setup calls
+	gomock.InOrder(
+		// Return that it was not found
+		driver.
+			EXPECT().
+			Inspect([]string{name}).
+			Return(nil, fmt.Errorf("not found")).
+			Times(1),
+
+		// Return too many volumes
```
Comment (piyush-nimbalkar) [inline pkg/util/volume_test.go]: nit: wrong comment

File: csi/controller_test.go [hunk 7]
```diff
@@ -912,6 +912,504 @@ func TestControllerListVolumes(t *testing.T) {
...
+		},
+	}
+	_, err = c.CreateVolume(context.Background(), req)
+	assert.NotNil(t, err)
+	serverError, ok = status.FromError(err)
+	assert.True(t, ok)
+	assert.Equal(t, serverError.Code(), codes.InvalidArgument)
+	assert.Contains(t, serverError.Message(), "CapacityRange")
+
+	// RequredBytes is zero
```
Comment (piyush-nimbalkar) [inline csi/controller_test.go]: nit: *required

File: csi/csi.go [hunk 8]
```diff
@@ -43,12 +44,13 @@ type OsdCsiServerConfig struct {
 // proxies CSI requests for a single specific driver
 type OsdCsiServer struct {
 	Server
-	listener net.Listener
-	server   *grpc.Server
-	driver   volume.VolumeDriver
-	wg       sync.WaitGroup
-	running  bool
-	lock     sync.Mutex
+	listener    net.Listener
```
Comment (sangleganesh) [inline csi/csi.go]: Please add comments for member.

File: csi/csi.go [hunk 9]
```diff
@@ -43,12 +44,13 @@ type OsdCsiServerConfig struct {
...
+	server      *grpc.Server
+	driver      volume.VolumeDriver
+	wg          sync.WaitGroup
+	running     bool
+	lock        sync.Mutex
```
Comment (sangleganesh) [inline csi/csi.go]: Does the lock protect access to all members ? 
If so embed the lock within the struct.
If not comments can explain what is being protected using this lock.

File: pkg/util/volume_test.go [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+	defer mc.Finish()
+
+	driver := mock.NewMockVolumeDriver(mc)
+
+	// Setup calls
+	name := "myvolume"
+	gomock.InOrder(
+		driver.
+			EXPECT().
+			Inspect([]string{name}).
```
Comment (sangleganesh) [inline pkg/util/volume_test.go]: Add test for Inspect(nil) and Inspect([]string{}) if not already.

File: csi/controller.go [hunk 11]
```diff
@@ -315,11 +316,154 @@ func osdVolumeAttributes(v *api.Volume) map[string]string {
...
 
+// CreateVolume is a CSI API which creates a volume on OSD
+// This function supports snapshots if the parent volume id is supplied
+// in the parameters.
+func (s *OsdCsiServer) CreateVolume(
+	ctx context.Context,
+	req *csi.CreateVolumeRequest,
+) (*csi.CreateVolumeResponse, error) {
+
+	// Log request
```
Comment (sangleganesh) [inline csi/controller.go]: Add nil check for req ?

File: csi/controller.go [hunk 12]
```diff
@@ -315,11 +316,154 @@ func osdVolumeAttributes(v *api.Volume) map[string]string {
...
+	resp := &csi.CreateVolumeResponse{
+		Reply: &csi.CreateVolumeResponse_Result_{
+			Result: &csi.CreateVolumeResponse_Result{
+				VolumeInfo: volume,
+			},
+		},
+	}
+
+	// Check if the volume has already been created or is in process of creation
+	v, err := util.VolumeFromName(s.driver, req.GetName())
```
Comment (sangleganesh) [inline csi/controller.go]: This looks like an error. 
Create must return error if already created - atleast that seems to be general behavior.

File: csi/controller.go [hunk 13]
```diff
@@ -315,11 +316,154 @@ func osdVolumeAttributes(v *api.Volume) map[string]string {
...
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume id must be provided")
+	}
+
+	err := s.driver.Delete(req.GetVolumeId())
+	if err != nil {
+		e := fmt.Sprintf("Unable to delete volume with id %s: %s",
+			req.GetVolumeId(),
+			err.Error())
+		dlog.Errorln(e)
```
Comment (sangleganesh) [inline csi/controller.go]: We should come up with guidelines on logging - automated schedulers sometimes keep sending invalid/stale requests and that can cause us to fill up logs.
In this case since the error is returned to the caller one can debug the issue - so the log does not help - so i would not log (my preference is to not log on invalid input, just like we do not log on version or volume name being empty). 
If not anything, for consistency purposes.

File: csi/controller.go [hunk 14]
```diff
@@ -315,11 +316,154 @@ func osdVolumeAttributes(v *api.Volume) map[string]string {
...
+	spec, locator, source, err := s.specHandler.SpecFromOpts(req.GetParameters())
+	if err != nil {
+		e := fmt.Sprintf("Unable to get parameters: %s\n", err.Error())
+		dlog.Errorln(e)
+		return nil, status.Error(codes.InvalidArgument, e)
+	}
+
+	// Check if the caller is asking to create a snapshot or for a new volume
+	var id string
+	if source != nil && len(source.GetParent()) != 0 {
```
Comment (sangleganesh) [inline csi/controller.go]: This should be left to the drivers to handle ?

File: csi/controller.go (hunk 11 above)
Comment (lpabon) [reply csi/controller.go]: afaik, grpc will not call with nil, it will not allow it.

File: csi/controller.go (hunk 12 above)
Comment (lpabon) [reply csi/controller.go]: If already created, it will return the information according to the spec.

File: csi/controller.go (hunk 14 above)
Comment (lpabon) [reply csi/controller.go]: That's how it works for the docker manager in osd.

File: csi/controller.go (hunk 13 above)
Comment (lpabon) [reply csi/controller.go]: I'll create an issue for this to follow up.

Comment (lpabon) [reply csi/controller.go]: Created #223

File: csi/csi.go (hunk 8 above)
Comment (lpabon) [reply csi/csi.go]: Not sure what you mean, could you show me an example? or are you saying we should add comments from now on for all osd structs?

File: csi/csi.go (hunk 9 above)
Comment (lpabon) [reply csi/csi.go]: locks access to start and stop only

File: pkg/util/volume_test.go (hunk 10 above)
Comment (lpabon) [reply pkg/util/volume_test.go]: Wouldn't that mean that someone has to pass nil into Inspect()? Could you show me an example of the code in CSI where that would happen?

Comment (sangleganesh) [reply pkg/util/volume_test.go]: If it can be done, it needs to be handled, regardless of someone is doing it today. Can you guarantee that no one in future will do it ?

File: csi/csi.go (hunk 8 above)
Comment (sangleganesh) [reply csi/csi.go]: https://github.com/libopenstorage/openstorage/blob/master/api/api.go

File: pkg/util/volume_test.go (hunk 10 above)
Comment (lpabon) [reply pkg/util/volume_test.go]: I think I am confused because I am not testing Inspect(), or expecting to have inspect be passed with nil. If the unit test passes nil to Inspect() then it will fail, since it expected something else. I think what you want to be tested is already being tested, right?

File: csi/csi.go (hunk 8 above)
Comment (lpabon) [reply csi/csi.go]: I see some done and some not. This is an internal struct, does it need to be commented in that manner? I agree with you that `OsdCsiServerConfig` should definitely have each member commented since it is used by the caller.

Comment (lpabon) [reply csi/csi.go]: Created #224
